articles/150_roslaunch.md

```diff
@@ -428,62 +428,168 @@ The remapping design document[^static_remapping] goes into detail on how you can
+
+This API could also have very low latency to launch nodes.
+However, there also is no way to get feedback about the success or failure of loaded nodes.
+STDOUT cannot be used because a composable node logging messages to STDOUT is assumed to be very common and would conflict.
+Since STDIN is always available, it would be possible to unload a node via this API.
+
+###### API using ROS Services or Topics
+
+Lastly, a container process API may be defined by ROS services or topics.
+
+Advantages
+* No config file parsing
+* Works with read-only file systems
+* Can indicate if a node was successfully loaded
+* Can create API to trigger launch events
+* Can tell if a container process supports this interface
+
+Disadvantages
+* Must wait for the service API to become available
+* Cannot stop dynamically loaded nodes from creating the same services
+
+This is the only option discussed which can communicate the success or failure of dynamically launched nodes.
+It is also the only option that allows introspection.
+However, this option has the highest potential delay from when the container process is spawned to when nodes may be loaded.
+
+##### Proposed Container process API
+
+This is a proposal for a container process API to be used by the launch system.
+
+###### Command Line Arguments
+The container process must accept command line arguments including log level, remapping, and parameters.
+These command line arguments must not be applied to dynamically launched nodes.
+The launch system will pass these arguments to the container process in the same way it would pass them to a node.
+If a remap rule would apply to a launch service, the launch system should try to use the remapped service name instead.
+
+###### ROS Services
+The container process must offer all of the following services.
+
+* `_launch/load_node`
+* `_launch/unload_node`
+* `_launch/list_nodes`
+
+The services are hidden to avoid colliding with user created services.
+`load_node` will be called by the launch system when a composable node is to be dynamically loaded, and `unload_node` destroys a composable node.
+`list_nodes` is not called by launch system, and is only provided for introspection.
+
+####### load_node
+
+```
+# A ROS package the composable node can be found in
+string package_name
+# a plugin within that package
+string plugin_name
+
+# Name the composable node should use
+string node_name
+# Namespace the composable node should use
+string namespace
+# Values from message rcl_interfaces/Log
+uint8 log_level
+# Remap rules
+# TODO(sloretz) rcl_interfaces message for remap rules?
+string[] remap_rules
+# Parameters to set
+rcl_interfaces/Parameter[] parameters
+
+# key/value arguments that are specific to a type of container process
+rcl_interfaces/Parameter extra_arguments
+---
+# True if the node was successfully loaded
+bool success
+# Name of the loaded composable node (including namespace)
+string full_node_name
```

Comment:
When is this value different from the concatenation of the two request fields `namespace` and `node_name`? If never, could it just be removed?

Reply:
Since multiple nodes can have the same `full_node_name` maybe returning a unique id here would be better and would allow to precisely specify which node to unload if the full name is not unique.

Reply:
If you make a way to optionally set the name and/or namespace, then the full name would be useful to know at this point (because depends on what is in the code).

There are already a few places where we assume unique node names (set/get parameter services for instance), so introducing a unique identifier here may not be the best thing, especially if we have to do a lot of work to generate one (call a UUID lib or something). If we just use the pointer to the node instance, then maybe that's ok, but the real solution is to address the node uniqueness design issue.

A potential compromise is that we continue to assume the node name is unique, but the container fails this service call if the new node's name is already present in the container.